File: android/app/src/main/java/com/babelwords/app/ads/ConsentManager.kt
package com.babelwords.com.ads

import android.app.Activity
import android.content.Context
import android.util.Log
import com.google.android.ump.ConsentForm
import com.google.android.ump.ConsentInformation
import com.google.android.ump.ConsentRequestParameters
import com.google.android.ump.FormError
import com.google.android.ump.UserMessagingPlatform

/**
 * UMP (User Messaging Platform) consent manager for GDPR/EEA compliance.
 *
 * Handles consent info collection, form display, and building ad requests with
 * the appropriate consent flags (personalized vs. non-personalized).
 *
 * Must be initialized before loading any ads. The consent flow runs asynchronously;
 * onConsentReady callback fires when consent is resolved (or consent form is
 * not needed for the user's region).
 *
 * Flow:
 *   1. requestConsentInfo() — updates consent status with Google's servers
 *   2. If a form is available → loadAndShowConsentFormIfRequired()
 *   3. onConsentReady(true) ↔ ad requests can proceed
 *      ↔ if consent is not obtained, ads will be requested with non-personalized flag
 *
 * The app does NOT block ads on consent — it adjusts the request personalization.
 * Unknown/declined consent → npa=1 (non-personalized) ads still serve.
 */
class ConsentManager(private val context: Context) {
    private val TAG = "ConsentManager"

    private val consentInformation: ConsentInformation by lazy {
        UserMessagingPlatform.getConsentInformation(context)
    }

    private var consentForm: ConsentForm? = null
    private var isProcessing = false

    /**
     * Request consent info update, then show the form if needed.
     * [onConsentReady] fires with true when consent is resolved (or not needed).
     * Even if consent is not obtained, you can still load ads (non-personalized).
     */
    fun requestConsent(activity: Activity, onConsentReady: (Boolean) -> Unit) {
        if (isProcessing) return
        isProcessing = true

        val params = ConsentRequestParameters.Builder()
            .setTagForUnderAgeOfConsent(false)
            .build()

        consentInformation.requestConsentInfoUpdate(
            activity,
            params,
            {
                // Consent info updated successfully
                isProcessing = false
                val canRequestAds = consentInformation.canRequestAds()
                Log.d(TAG, "Consent info updated. canRequestAds=$canRequestAds, formAvailable=${consentInformation.isConsentFormAvailable}")

                if (consentInformation.isConsentFormAvailable) {
                    loadAndShowConsentForm(activity, onConsentReady)
                } else {
                    onConsentReady(true)
                }
            },
            { requestConsentError: FormError ->
                isProcessing = false
                Log.w(TAG, "Consent info update failed: ${requestConsentError.message}")
                // Allow ads anyway on consent failure (non-personalized fallback)
                onConsentReady(true)
            }
        )
    }

    private fun loadAndShowConsentForm(activity: Activity, onConsentReady: (Boolean) -> Unit) {
        UserMessagingPlatform.loadConsentForm(
            activity,
            { form: ConsentForm ->
                consentForm = form
                val consentStatus = consentInformation.consentStatus
                if (consentStatus == ConsentInformation.ConsentStatus.REQUIRED) {
                    form.show(activity) { formError: FormError? ->
                        if (formError != null) {
                            Log.w(TAG, "Consent form show error: ${formError.message}")
                        }
                        Log.d(TAG, "Consent form dismissed. canRequestAds=${consentInformation.canRequestAds()}")
                        onConsentReady(consentInformation.canRequestAds())
                    }
                } else {
                    // Form available but consent already obtained (or not needed)
                    onConsentReady(true)
                }
            },
            { formError: FormError ->
                Log.w(TAG, "Consent form load failed: ${formError.message}")
                onConsentReady(true)
            }
        )
    }

    /**
     * Reset consent for testing. Call this only in debug builds.
     */
    fun resetConsent() {
        consentInformation.reset()
    }

    /**
     * Build an AdRequest with appropriate consent flags.
     * Returns the request; the caller passes it to AdMob.load().
     */
    fun buildAdRequest(): com.google.android.gms.ads.AdRequest {
        val canRequestAds = consentInformation.canRequestAds()
        val consentStatus = consentInformation.consentStatus

        // UMP SDK 3.x + play-services-ads 24.x automatically applies consent flags
        // to all AdRequests via the stored consent string. Manual npa=1 bundle is
        // no longer needed (and AdMobAdapter class is unavailable in 24.x).
        return if (canRequestAds && consentStatus == ConsentInformation.ConsentStatus.OBTAINED) {
            Log.d(TAG, "Building ad request (consent obtained)")
            com.google.android.gms.ads.AdRequest.Builder().build()
        } else {
            Log.d(TAG, "Building ad request (non-personalized via UMP consent string)")
            com.google.android.gms.ads.AdRequest.Builder().build()
        }
    }

    fun isConsentAvailable(): Boolean = consentInformation.isConsentFormAvailable
}

File: ios/BabelWords/Managers/ConsentManager.swift
import Foundation
import GoogleMobileAds
@preconcurrency import UserMessagingPlatform

// MARK: - Testability protocols

/// Abstracts `UMPConsentInformation` so tests can inject a mock without hitting
/// the live UMP SDK.
protocol ConsentInformationProviding: AnyObject {
    func requestConsentInfoUpdate(
        with parameters: UMPRequestParameters?,
        completionHandler: @escaping @Sendable (Error?) -> Void
    )
    var canRequestAds: Bool { get }
    var formStatus: UMPFormStatus { get }
    var consentStatus: UMPConsentStatus { get }
    func reset()
}

/// Abstracts `UMPConsentForm` presentation so tests can inject a mock form.
protocol ConsentFormPresenting: AnyObject {
    func present(
        from viewController: UIViewController?,
        completionHandler: (@Sendable (Error?) -> Void)?
    )
}

/// Closure type that loads a consent form and calls back with a
/// `ConsentFormPresenting` (or an error).  In production this wraps
/// `UMPConsentForm.load`; in tests a stub closure is injected.
typealias ConsentFormLoader = (@escaping @Sendable (ConsentFormPresenting?, Error?) -> Void) -> Void

// Production conformances (no additional implementation needed).
extension UMPConsentInformation: ConsentInformationProviding {}
extension UMPConsentForm: ConsentFormPresenting {}

// MARK: -

/// UMP (User Messaging Platform) consent manager for GDPR/EEA compliance.
/// Mirrors the Android `ConsentManager` behavior.
///
/// ## Thread safety
/// This class is main-thread confined. UMP completion closures remain
/// nonisolated at the SDK boundary because the SDK does not accept actor-isolated
/// function values; their state/UI work is normalized onto the main queue below.
///
/// The `@preconcurrency import UserMessagingPlatform` suppresses Swift 6
/// "Sendable" diagnostics on UMP closure types that the SDK has not yet annotated.
///
/// ## Off-main-thread guard
/// SDK callbacks that arrive off-main complete with payload-free recovery; the
/// non-Sendable form is never transferred into a detached actor task.
final class ConsentManager: NSObject, @unchecked Sendable {
    private let TAG = "ConsentManager"

    // MARK: - Dependencies (injectable for testing)

    private let consentInformation: ConsentInformationProviding
    private let formLoader: ConsentFormLoader

    // MARK: - State

    private var isProcessing = false
    private var pendingCallbacks: [(Bool) -> Void] = []
    private var pendingViewController: UIViewController?

    // MARK: - Consent-change notification

    /// Assigned by the owner (e.g. MainViewController) to receive a notification
    /// whenever consent completes.  The canonical use-case is invalidating
    /// any ad cache that was built with the previous consent signal.
    var onConsentUpdated: ((Bool) -> Void)?

    // MARK: - Initializers

    /// Production initializer — uses the live UMP SDK.
    override init() {
        self.consentInformation = UMPConsentInformation.sharedInstance
        self.formLoader = { completion in
            UMPConsentForm.load { form, error in
                completion(form, error)
            }
        }
        super.init()
    }

    /// Testing initializer — allows injecting mock consent information and a
    /// stub form loader so the full request → load → present → completeConsent
    /// path can be exercised without hitting the live UMP SDK.
    init(
        consentInformation: ConsentInformationProviding,
        formLoader: @escaping ConsentFormLoader
    ) {
        self.consentInformation = consentInformation
        self.formLoader = formLoader
        super.init()
    }

    // MARK: - Public API

    /// Request consent info, show the form if required, and call back when ads can proceed.
    func requestConsent(from viewController: UIViewController, onConsentReady: @escaping (Bool) -> Void) {
        precondition(Thread.isMainThread, "ConsentManager must be used on the main thread")
        pendingCallbacks.append(onConsentReady)
        guard !isProcessing else { return }
        isProcessing = true
        pendingViewController = viewController

        let parameters: UMPRequestParameters? = {
            let parameters = UMPRequestParameters()
            parameters.tagForUnderAgeOfConsent = false
            return parameters
        }()

        consentInformation.requestConsentInfoUpdate(with: parameters) { [weak self] error in
            let errorDescription = error?.localizedDescription
            guard Thread.isMainThread else {
                DispatchQueue.main.async { [weak self] in
                    self?.handleConsentInfoUpdateResult(errorDescription: errorDescription)
                }
                return
            }
            self?.handleConsentInfoUpdateResult(errorDescription: errorDescription)
        }
    }

    // MARK: - Consent-info update result

    private func handleConsentInfoUpdateResult(
        errorDescription: String?
    ) {
        if let errorDescription = errorDescription {
            print("[\(TAG)] Consent info update failed: \(errorDescription)")
            completeConsent(consentInformation.canRequestAds)
            return
        }

        let canRequestAds = consentInformation.canRequestAds
        print("[\(TAG)] Consent info updated. canRequestAds=\(canRequestAds), formAvailable=\(consentInformation.formStatus == .available)")

        if consentInformation.formStatus == .available {
            guard let viewController = pendingViewController else {
                completeConsent(canRequestAds)
                return
            }
            loadAndShowConsentForm(from: viewController)
        } else {
            completeConsent(canRequestAds)
        }
    }

    // MARK: - Form load + present

    private func loadAndShowConsentForm(from viewController: UIViewController) {
        formLoader { [weak self] form, error in
            let errorDescription = error?.localizedDescription
            if Thread.isMainThread {
                self?.handleFormLoadResult(
                    form: form,
                    errorDescription: errorDescription,
                    viewController: viewController
                )
            } else {
                DispatchQueue.main.async { [weak self] in
                    guard let self = self else { return }
                    print("[\(self.TAG)] Consent form loaded off-main; completing without presenting")
                    self.completeConsent(self.consentInformation.canRequestAds)
                }
            }
        }
    }

    private func handleFormLoadResult(
        form: ConsentFormPresenting?,
        errorDescription: String?,
        viewController: UIViewController
    ) {
        if let errorDescription = errorDescription {
            print("[\(TAG)] Consent form load failed: \(errorDescription)")
            completeConsent(consentInformation.canRequestAds)
            return
        }

        guard let form = form else {
            completeConsent(consentInformation.canRequestAds)
            return
        }

        if consentInformation.consentStatus == .required {
            form.present(from: viewController) { [weak self] error in
                let errorDescription = error?.localizedDescription
                guard Thread.isMainThread else {
                    DispatchQueue.main.async { [weak self] in
                        self?.handleFormPresentResult(errorDescription: errorDescription)
                    }
                    return
                }
                self?.handleFormPresentResult(errorDescription: errorDescription)
            }
        } else {
            completeConsent(consentInformation.canRequestAds)
        }
    }

    private func handleFormPresentResult(errorDescription: String?) {
        if let errorDescription = errorDescription {
            print("[\(TAG)] Consent form show error: \(errorDescription)")
        }
        print("[\(TAG)] Consent form dismissed. canRequestAds=\(consentInformation.canRequestAds)")
        completeConsent(consentInformation.canRequestAds)
    }

    // MARK: - Complete

    private func completeConsent(_ canRequestAds: Bool) {
        isProcessing = false
        pendingViewController = nil
        let callbacks = pendingCallbacks
        pendingCallbacks.removeAll()
        callbacks.forEach { $0(canRequestAds) }
        onConsentUpdated?(canRequestAds)
    }

    // MARK: - Helpers

    /// Build a GADRequest with the appropriate consent context.
    func buildAdRequest() -> GADRequest {
        let request = GADRequest()
        // UMP consent string is automatically applied by the Google Mobile Ads SDK.
        return request
    }

    func resetConsent() {
        precondition(Thread.isMainThread, "ConsentManager must be used on the main thread")
        consentInformation.reset()
    }

    func isConsentAvailable() -> Bool {
        consentInformation.formStatus == .available
    }
}
